### ISSUE TITLE
Write tool which can convert back to PO

### ISSUE THREAD

**Issue by mgeisler** [google/mdbook-i18n-helpers#30]
This idea is from https://github.com/rust-embedded/book/issues/326: we should write a converter tool which takes two Markdown files as input and outputs a PO file.

More concretely, the tool should take an `en/foo.md` and `xx/foo.md` file and output a `xx.po` file. The tool will call `extract_messages` on both files and line up the results. It will use the messages from `en/foo.md` as the `msgid` and the corresponding message from `foo/xx.md` as the `msgstr`.

The output is marked fuzzy to ensure that a human translator double-checks it all before publication.

**Comment by memark** [comment]
I'd be happy to take a look at this!

Are there any good example .md files?

**Comment by mgeisler** [comment]
Hejsa! Thanks for looking at this!

I was thinking to use some of the [Rust book translations](https://doc.rust-lang.org/book/appendix-06-translation.html) to start with, e.g., pick one here which is somewhat current and which you can read :slightly_smiling_face: 

The [Rust Embedded book](https://docs.rust-embedded.org/book/#translations) is also translated into a few languages — that's where the idea came from.

You might want be able to reuse some of the logic in [`mdbook-i18n-normalize`](https://github.com/google/mdbook-i18n-helpers/blob/main/src/bin/mdbook-i18n-normalize.rs). That tool takes a PO file looks at each `(msgid, msgstr)` pair. It then runs `extract_messages` on both to get two lists:

```rust
let msgids = ["msgid_1", "msgid_2", "msgid_1"];
let msgstrs = ["msgstr_1", "msgstr_2", "msgstr_1"];
```

Finally, it zips those lists and output the `("msgid_N", "msgstr_N")` pairs to the PO file. Having written this, it occurs to me that this is what your tool will have to do as well.

One feature your tool could have would be to (attempt to) synchronize the translations on various points. So if the source text is

```markdown
# The Foo Project

Welcome to the Foo projects...

## Getting Started

To install Foo, do ...
```

and the translated document is

```markdown
# Foo-projektet

Velkommen til Foo-projektet...

## Komme godt igang

For at installere Foo, ...
```

then the conversion tool could align on the `#` and `##` headings. This would make it robust against extra paragraphs being inserted in either language.

Similar with lists and block quotes, they could perhaps also be synchronization points. This is just an idea, I'm not sure it's useful :smile:

**Comment by mgeisler** [comment]
> It would be a separate bin crate then?

I missed this part, I think it could just be a new binary next to the others to start with. If people like the tool but somehow don't want to see `mdbook-*` binaries on their system, then we could consider moving this binary to its own crate. Luckily, mdbook-i18n-helpers is already a library crate, so it's easy to depend on it from other crates.

**Comment by memark** [comment]
>I think it could just be a new binary next to the others to start with.

Yes, that's what I mean by "separate bin crate". I agree that we might do a separate "package" later on if needed.

Additionally, I'm a bit held up at the moment with other work, but I'll keep working in this whenever I have the time.

>I was thinking to use some of the [Rust book translations](https://doc.rust-lang.org/book/appendix-06-translation.html) to start with
From that link, how do I get the actual translated .md file? It seems that all that's available is either .po or .html files.

**Comment by mgeisler** [comment]
Hejsa!

> > I think it could just be a new binary next to the others to start with.
> 
> Yes, that's what I mean by "separate bin crate". I agree that we might do a separate "package" later on if needed.

I see, great!
 
> Additionally, I'm a bit held up at the moment with other work, but I'll keep working in this whenever I have the time.

That's completely fine, I know the feeling :smile: 

> > I was thinking to use some of the [Rust book translations](https://doc.rust-lang.org/book/appendix-06-translation.html) to start with
> 
> From that link, how do I get the actual translated .md file? It seems that all that's available is either .po or .html files.

Clicking through to the Swedish translation (to keep a Scandinavian theme — I'm from Denmark and I guess you're from Sweden), leads me to https://github.com/sebras/book/tree/master/src where I see a bunch of translated files. So I would use these two files an a starting point:

- https://github.com/sebras/book/blob/master/src/ch00-00-introduction.md
- https://github.com/rust-lang/book/blob/main/src/ch00-00-introduction.md

I'm not sure if the idea actually works... will 10% of the paragraphs line up, or 80%? Just from skimming those two files, it looks like a lot of text lines up very nicely.

**Comment by memark** [comment]
Hejsan!

I've (finally) started looking into this. The code required is pretty straight forward for reading, parsing and extracting. The hard part is the syncing. Having looked at a couple different translations of the book, they can differ a lot, a little, or not at all.

Some examples.
Filename: number-of-extracted-messages

ch00-00-introduction.en.md: 45
ch00-00-introduction.sv.md: 46

ch01-02-hello-world.en.md: 45
ch01-02-hello-world.sv.md: 42

ch02-00-guessing-game-tutorial.en.md: 177
ch02-00-guessing-game-tutorial.pt.md: 252

Trying to line them up might work in really simple scenarios, where the structure is still intact. But in the PT case above, the EN version is really far ahead. The main problem I see is that there is a lack of context in the MD files. If a single heading (or other sync point of choice) is left out in either file, the remaining message pairs will be garbage. Even with a small difference like 42 vs 45, there is no way to tell if the diff was at message 1 or 40.

So what is "good enough" here? Should we fail as soon as something appears to be off? Or greedily try to do what we can? What's the threshold? Strict or lax? Need some guidance here.

**Comment by memark** [comment]
Btw, having run just my own tool (which calls `extract_messages()`) a number of times, closing down my editor and my terminal, I get a number of zoombie processes that linger consuming CPU. (I only noticed this because this is the first time ever the fan on my M3 has started. :)) Is this expected?

<img width="618" alt="image" src="https://github.com/google/mdbook-i18n-helpers/assets/318504/1fdaf003-02ed-4750-85bc-78295ab0e4e1">

**Comment by mgeisler** [comment]
> Trying to line them up might work in really simple scenarios, where the structure is still intact. But in the PT case above, the EN version is really far ahead.

That is a very good point and I don't think there is a clear right answer here.

Our `mdbook-i18n-normalize` tool has a similar problem: it normalizes a `.po` file by re-applying the extraction logic. The idea is that we've improved the way we extract messages over time and so this tool lets you migrate an old `xx.po` file to benefit from the latest improvements. Biggest example of this: we used to extract a Markdown table as a single unit, now we extract the text cell by cell.

When we normalize each `(msgid, msgstr)` pair, we run into similar problems as what you have here: we could easily end up with different number of extracted messages from each of `msgid` and `msgstr`. The normalization tool handles this in a very primitive way here:

https://github.com/google/mdbook-i18n-helpers/blob/2b14491962254ec2ff3f12ffeb59f09f7f4686ae/i18n-helpers/src/normalize.rs#L203-L220

If there are extra messages in the `msgstr` field, it accumulates them in the final `msgid` field. This is confusing in its own way and indirectly lead to #125.

> The main problem I see is that there is a lack of context in the MD files. If a single heading (or other sync point of choice) is left out in either file, the remaining message pairs will be garbage. Even with a small difference like 42 vs 45, there is no way to tell if the diff was at message 1 or 40.

You mention a "sync point"... which is something we haven't looked into before! For the use case of generating a `.po` file from two `.md` files, I think establishing some sync points could be very helpful. This could be Markdown features which we hope are stable over time:

- headings of different levels
- lists
- tables
- block quote

and perhaps others?

I haven't thought this through, but I could imagine using a [largest common subsequence or diffing algorithm](https://florian.github.io/diffing/) to find common parts of the Markdown AST between the two documents.

The flow would be something like this:

1. Extract the AST for each document
2. Normalize all text so that you're left with a normalized outline. Something like

   ```markdown

   # Heading 1
   ## Heading 1.1

   Paragraph 1 in 1.1

   Paragraph 2 in 1.1

   - Item 1 in 1.1
   - Item 2 in 1.1

   Paragraph 3 in 1.1
   ```

   Here I'm thinking that it'll be good to include the heading numbers in the normalized paragraph and list item text to give the diffing algorithm something to "attach to"?

3. Diff the two normalized documents. Everything that matches gives you a mapping from `msgid` (from the first document) to `msgstr` (the second document). Things that are extra in the second document can be dropped, things that are missing in the second document just have a `msgstr` equal to `""`
4. Go back from the normalized outlines to the original documents and output a `.po` file.

At the end of the day, one document will be the source. So regardless of what the translated document contains, the source will win: if there are 10 paragraphs in the source document, then there should also be 10 paragraphs in the translation. This is kind of a built in limitation of the way we attempt to chop up the source document when we do translations.

**Comment by mgeisler** [comment]
> Btw, having run just my own tool (which calls `extract_messages()`) a number of times, closing down my editor and my terminal, I get a number of zoombie processes that linger consuming CPU.
> (I only noticed this because this is the first time ever the fan on my M3 has started.)
> Is this expected?

Nope, I would not expect this — everything in this library should be completely in-process and not spawn anything. It only does text to text transformations.

The `group_events-...` name is interesting: we don't have a binary with that name, so I guess it comes from somewhere else?

**Comment by mgeisler** [comment]
Hi @memark, let me unassign this for now — we can assign it back to you or someone else later when activity picks up.

**Comment by michael-kerscher** [comment]
> I haven't thought this through, but I could imagine using a [largest common subsequence or diffing algorithm](https://florian.github.io/diffing/) to find common parts of the Markdown AST between the two documents.

## Overall idea
When thinking about this I also think this is best done by matching on the AST.
Lets parse the markdown with pulldown_cmark and use the [`Event`](https://docs.rs/pulldown-cmark/latest/pulldown_cmark/enum.Event.html)s as symbols without the content. E.g. `Text("foobar")` just becomes `Text`, `Start(Tag)` and `End(TagEnd)` have important structural information in [`Tag`](https://docs.rs/pulldown-cmark/latest/pulldown_cmark/enum.Tag.html) and [`TagEnd`](https://docs.rs/pulldown-cmark/latest/pulldown_cmark/enum.TagEnd.html) and are kept. All elements that are structural informations are kept.

The sequence should still store enough information in such a way that we can recover the original document.

Now we want to get the best matching between the two structural sequences.

## Algorithms
Let's compare some algorithms. Bioinformatics has to deal a lot with sequence matching, so this is a good inspiration 
- [Longest Common Subsequence](https://en.wikipedia.org/wiki/Longest_common_subsequence) (LCS): Finds the longest common subsequence that appears in the same order in both sequences. This would provide us with all elements that are actually equal and makes it easy to dismiss the "extra" bits. As far as I understand, it only matches unmodified bits of structure. If someone changes an element to a different type this will be missed.
- [Needleman-Wunsch](https://en.wikipedia.org/wiki/Needleman%E2%80%93Wunsch_algorithm): Performs global alignment over the entire length to find the best overall match. This is usually used if two sequences are similar and of roughly equal size. This would just match up the structure as good as possible and could deal with the issue of translators changing the structure to a different type.
- [Smit-Waterman](https://en.wikipedia.org/wiki/Smith%E2%80%93Waterman_algorithm): Performs local alignment that identifies regions of high similarity within the sequences. Usually this is used for dissimilar sequences that are suspected to contain regions of similarity. Not the best fit for our use case.

Let's think about this for a minute:

The Longest Common Subsequence can exactly provide what we need. It would try to match the structure and all structures that match are returned and this has with high probability (but not 1.0) matching content that was translated. It discards every changed structural element (`[A B C]` → `[A D C]`) but would keep the `[A C]` elements. In reality a lot of algorithms provide a vector of changes so we could also see these inserts, deletions.

The last two could also be candidates for the problem we face. From these I would propose the global [alignment](https://en.wikipedia.org/wiki/Sequence_alignment#Global_and_local_alignments) as by definition, the second document is a document that was once the exact same structure but differed (ideally) only by the non-structural text that is invisible to the algorithm. This would allow to try to match (`[A B C]` → `[A D C]`) in a way that it recognized that `B` was *edited* into `D`. Which might be wrong but could be true.

I don't think it really matters for our usecase if we use a LCS or a Needleman-Wunsch and they both should provide similar output. The later just seems to be a bit more flexible in the sequence comparison.

## Crates:
### Longest common Subsequence
[crates](https://crates.io/search?q=longest%20common%20subsequence)
- [lcs_rs](https://crates.io/crates/lcs_rs) (v0.1.1): API requires strings, does not return changes
- [slice-diff-patch](https://crates.io/crates/slice-diff-patch) (v1.2.4): works with parametrized vectors of elements (supporting `PartialEq`. [provides change vector](https://docs.rs/slice-diff-patch/1.2.4/slice_diff_patch/) (insert, remove, update,...). Looks very promising
- [diffs](https://crates.io/crates/diffs) (v0.5.1): This supports a Binary, a Meyers (used for git) and a Patience (git also supports this, more human readable diff). This might be usable but the documentation is pretty slim, has [tests as examples](https://nest.pijul.com/pijul/diffs:main/XNQ4AFHIXFKIQ.CEAAA) and might support our use case.


### Needleman-Wunsch
[crates](https://crates.io/search?q=needleman-wunsch):
- [stringzilla](https://crates.io/crates/stringzilla) (v3.12.5): heavily optimized string matching crate, leveraging SIMD. using [`[u8]` in the API](https://docs.rs/stringzilla/3.12.5/stringzilla/sz/fn.alignment_score.html). While this might be usable if our structural elements are translated to numbers, I don't see how I can get the alignment operations as a result.
- [needleman](https://crates.io/crates/needleman) (v0.2.0): API expects two strings and I don't see how I can get the alignment as result.
- [seal](https://crates.io/crates/seal) (v0.1.5, last release 3 years ago...): This is a generic API that just needs a (equality) compare function between elements of both sequences. And it supports both local and global alignment, even if I don't think local would give as any advantage but this is easily testable.

### What crate will I use?
I will pick the LCS supporting crate [slice-diff-patch](https://crates.io/crates/slice-diff-patch) (v1.2.4) because the API supports parameterized vector elements, it provides a change vector and there was a recently release.

My next favorite option would have been [seal](https://crates.io/crates/seal) (v0.1.5) which supports Needleman-Wunsch and also supports parameterized vector elements, provides a change vector. But its last release is 3 years ago and has a low version number, indicating not much support.


## Challenges:
The problem gets difficult once there are extra regions because we don't know which `B` of the sequence `ABBA`, `ABBBA` was introduced newly. We could try to provide some additional hints for that case by using the text length of this structure as an additional property. `A(_) B(20) B(10) A(_)` and `A(_) B(23) B(50) B(8) A(_)` would lead to the conclusion that B(50) was added in the second structure.

## Ideas for encoding the structure to deal with structural changes:
- Structural Elements are encoded into Enum Variants and we dive as deep as necessary into the elements and keep all structural information
- Text is encoded by substituting each word with a new Enum Variant `Word` (or `Letter`, `Sentence`) under the assumption that translations use roughly the same amount of words but different amount of letters per word (and differently ordered). Some testing / benchmarking might provide us with data. The existing po files are the perfect source for generating data if `Letter`, `Word` (split by whitespace) or even rougher as `Sentence` (split by dot - or more generic non-whitespace separators) are better (because each sentence might be translated to exactly once sentence, providing perfect matching opportunities). Paragraphs are already encoded in the [Tag](https://docs.rs/pulldown-cmark/latest/pulldown_cmark/enum.Tag.html) but might otherwise also be an opportunity.

**Comment by michael-kerscher** [comment]
I quickly tested the matching approach and with the `lcs_diff` I can properly match up the [german translation](https://rust-lang-de.github.io/rustbook-de/ch00-00-introduction.html) of the [introduction](https://doc.rust-lang.org/book/ch00-00-introduction.html) of the rust lang book.

I used this page as test because it is long and has the challenge of a Note section in the beginning. This can be properly detected. But there is some merging of remove + insert operations into an update operation that is not working properly for code + softbreak situations. These happen when there is a code block at the end of a line followed by a softbreak.

A solution would be to override this behavior or even better use the underlying function [lcs_diff::diff()](https://docs.rs/lcs-diff/latest/lcs_diff/fn.diff.html) directly. The patch function that is offered by the [slice-diff-patch](https://crates.io/crates/slice-diff-patch) crate is not worth the additional trouble.

I need to clean my draft code up a bit, add the extract messages part and create the final catalog

**Comment by michael-kerscher** [comment]
Note: LCS seems to work well with specific cases but overall a Needleman-Wunsch algorithm might produce better results.

e.g. in the case:
- Heading, Text(A, 10 Sentences), Text(B, 20 Sentences), Text(C, 30 Sentences)
- Heading, Text(B, 20 Sentences), Text(C, 30 Sentences)

This might produce a matching of A and B and C is missed.

**Comment by michael-kerscher** [comment]
It turns out it is difficult to work with cmark event streams. While this is not entirely bad, there is still a lot of tuning necessary.

Another thought might be to use AI for this (oh no, he said AI! ;) ). Not in a "just give it two markdown files and create the po" prompt style but actually compare embeddings of (parts) of the original and translation. Then the similar embeddings should point us to the right pairs of translated paragraphs
